pulsar-broker/src/main/java/org/apache/pulsar/broker/MessagingServiceShutdownHook.java

```diff
@@ -59,9 +59,14 @@ public void run() {
 
             executor.execute(() -> {
                 try {
-                    service.close();
-                    future.complete(null);
-                } catch (PulsarServerException e) {
+                    service.closeAsync().whenComplete((result, throwable) -> {
+                        if (throwable != null) {
+                            future.completeExceptionally(throwable);
+                        } else {
+                            future.complete(result);
+                        }
+                    });
+                } catch (Exception e) {
```

Comment:
from this code it is difficult to understand why we have to catch a generic Exception.
we are calling a `closeAsync()` and we are handing errors in the `whenComplete` callback.

Reply:
I agree that it might be difficult to understand. What do you suggest?

Reply:
To not throw any Exception in closeAsync and remove the "throws PulsarServerException" clause

Reply:
That change wouldn't simplify this code in MessagingServiceShutdownHook. 
Unless the executor solution is modified, it's not possible get rid of the `.closeAsync().whenComplete` code block.

Reply:
Sorry, I don't want to be so picky, but  "catch (Exception e)" is really something we do not want to have if not strictly necessary..

It totally fine to me to handle the result with `whenComplete`, this is expected when you deal with CompletableFuture.
My problem is "catch Exception" around a method call to something that returns a CompletableFuture.

the only way to get rid of the "catch Exception" is to remove the  "throws PulsarServerException" clause from closeAsync.

`closeAsync` is a new method, so we can define the signature without the "throws" clause

Reply:
I guess we discussed it that catching all exceptions is fine?

Reply:
I changed it to be "catch (RuntimeException e) {" . It's necessary to catch possible runtime exceptions. Since there's no more checked exceptions in the most recent changes, it's sufficient to catch RuntimeExceptions. Are you fine with this?

Reply:
I put it back to "catch (Exception e) {" based on @merlimat's comment about handle all exceptions (that was in another code location,  https://github.com/apache/pulsar/pull/9308#discussion_r602616184 ).
Catching RuntimeException is actually not sufficient on Java because there are solutions like Lombok's SneakyThrows which allows to throw checked exceptions. That's why it's better to simply catch all Exceptions.